Added feature_importance.rst (#1391)
deepchecks/deepchecks: docs/source/user-guide/tabular/feature_importance.rst

.. _feature_importance:

====================
Feature Importance
====================

> 

Author: Done

link to feature drift
link to "how to make faster"

What is Feature Importance?
===========================
Feature importance is a ranking that represents the significance of input features to the model's predictions.
A feature with higher importance has more influence on the prediction of the model.
Feature importance can be general (meaning, for all model predictions, on average) or local (meaning, for a specific
sample).
There are many ways to calculate feature importance, some are generic for all models (such as Shapley values) and some
are specific for a specific model (such as the Gini index for decision trees).

Why Does Deepchecks Use Feature Importance?
===========================================
Deepchecks uses your model's feature importance for 2 main reasons:
* Help you find issues with your model or data, as in the check UnusedFeatures
* Display the most relevant information the check has found (for instance, if deepchecks found drift in many features,
as in the check TrainTestFeatureDrift, it would only display the features with the highest importance)

Note
-----
For most checks, deepchecks does not *require* the usage of feature importance, and that you can

> shorten or  :ref:`even skip <_feature_importance__calculation_too_long>`. this phase of the calculation. See

Author: Right

How Does Deepchecks Get Feature Importance?
===========================================
There are 3 ways in which deepchecks can get your model's feature importance:

Your Model Has a Built-in Feature Importance
--------------------------------------------
First of all, deepchecks searches for your model's built-in feature importance, as some scikit-learn models have.
Deepchecks looks for the attribute ``feature_importances_`` or ``coef_`` and uses that information if it exists.

You Insert Your Own Feature Importance Data
-------------------------------------------

> This can be done by using the ``features_importance`` input received by the ``run`` function, available in all

Author: Changed to "parameter" (and in other places)

checks and suites.
Deepchecks expects this data to be a ``pandas.Series`` where the index is feature names and the value is the calculated
importance.

Deepchecks Calculates the Feature Importance for You
----------------------------------------------------
If there's no built-in feature importance in the model or the user has not supplied feature importance data of their
own, deepchecks will calculate feature importance using scikit-learn's `permutation_importance <https://scikit-learn.org/stable/modules/generated/sklearn.inspection.permutation_importance.html>`_.

You can also force this action by using the ``feature_importance_force_permutation`` input received by the ``run``
function, available in all checks and suites.

.. _feature_importance__calculation_too_long:
What if the Feature Importance Calculation Takes Too Long?
----------------------------------------------------------

> Feature importance is a complex calculation which can take a lot of time, depending on the number of features and

Reviewer: So the method to skip FI calculation is to set it to 0? If so, say it explicitly

Author: Done

samples in your data.
However, besides for cetrain checks, deepchecks does not require feature importance.
Therefore, if you want deepchecks to skip the calculation of feature importance, you can use the

> ``feature_importance_timeout`` input received by the ``run`` function, available in all checks and suites. If

Reviewer: ```suggestion
``feature_importance_timeout`` argument of the ``run`` function, available in all checks and suites. If
```

sample).
There are many ways to calculate feature importance, some are generic for all models (such as `Shapley values <https://christophm.github.io/interpretable-ml-book/shapley.html>`_)
and some are specific for a specific model type (such as the `Gini importance <https://medium.com/the-artificial-impostor/feature-importance-measures-for-tree-models-part-i-47f187c1a2c3#:~:text=Gini%20Importance%20or%20Mean%20Decrease%20in%20Impurity%20(MDI)%20calculates%20each,number%20of%20samples%20it%20splits.>`_
for decision trees).

Why Does Deepchecks Use Feature Importance?
===========================================
Deepchecks uses your model's feature importance for 2 main reasons:
* Help you find issues with your model or data, as in the check :doc:`UnusedFeatures </checks_gallery/tabular/methodology/unused_features>`
* Display the most relevant information the check has found (for instance, if deepchecks found drift in many features,
as in the check :doc:`TrainTestFeatureDrift </checks_gallery/tabular/distribution/train_test_feature_drift>`,
it would only display the features with the highest importance)

> Note

Reviewer 2: try with .. note: instead of a title

checks and suites.
Deepchecks expects this data to be a ``pandas.Series`` where the index is feature names and the value is the calculated
importance.

>>> check = UnusedFeatures()
>>> check.run(ds_train, ds_test, model, feature_importance=pd.Series({'feat1': 0.3, 'feat2': 0.7}))

Deepchecks Calculates the Feature Importance for You
----------------------------------------------------
If there's no built-in feature importance in the model or the user has not supplied feature importance data of their
own, deepchecks will calculate feature importance using scikit-learn's `permutation_importance <https://scikit-learn.org/stable/modules/generated/sklearn.inspection.permutation_importance.html>`_.

You can also force this action by using the ``feature_importance_force_permutation`` parameter in the ``run``
function, available in all checks and suites.

>>> check = TrainTestFeatureDrift()
>>> check.run(ds_train, ds_test, model, feature_importance_force_permutation=True)

What if the Feature Importance Calculation Takes Too Long?
=========================================================

> Feature importance is a complex calculation which can take a lot of time, depending on the number of features and

Reviewer 2: ```suggestion
Permutation importance is a complex calculation which can take a lot of time, depending on the number of features and
```

samples in your data.
However, besides for cetrain checks, deepchecks does not require feature importance.
Therefore, if you want deepchecks to skip the calculation of feature importance, you can use the
``feature_importance_timeout`` parameter in the ``run`` function, available in all checks and suites. If
deepchecks projects the feature importance calculation to take more than this parameter, the process will be skipped.

> Configuring this parameter to 0 will ensure the calculation is always skipped.

Reviewer 2: I think the timeout thing is not clear enough
